### DISCUSSION TITLE
[2023-05] RP2040 Zero and custom keyboard preparations

### DISCUSSION THREAD

**Discussion by KenVanHoeylandt** [ByteWelder/Decktility#3] [Announcements]
I've started doing preparation work for a custom keyboard. It'll need it's own MCU, so I've started by replacing the Arduino Nano with an RP2040 Zero to create space. Some benefits of the RP2040 Zero:

- It's smaller
- It uses less power at 50 MHz than the Nano does at 16 MHz... and about the same at 133 MHz. (this was measured with no peripherals attached, using a USB power meter)
- It is a lot more powerful (MHz, RAM, ROM)
- Has an extra I2C interface

There are less pins available, but we don't even use half of them as-is.

The BlueTooth keyboard that is currently in use isn't very reliable when typing, due to the way the buttons work. I intend to make a custom one with micro switches a QMK firmware. A few weeks ago, I bought a box with about a dozen different micro switches, to find the one(s) that feel the best and have the quietest action.

One of the other changes that I'm planning is to use a power FET that has a smaller footprint. I would like to use 2 FETs: one for the main hardware (Pi + keyboard) and one for the fan. That way, I can turn on the fan while charging without turning on the Pi.

I also want the ability to power off the device entirely after Linux is done. This is currently not possible and requires the user to flip the switch after the shut off sequence. Much like old computers used to power down. I'm thinking of using a FET for this. If this goes to plan, I'll need to replace the power switch with a button. I'll use it in a way that you can make a relay keep on powering itself after an initial power on. I'll also have to think about a way to do a force shutdown, like the long press that you see on most devices.

Here's a preview for the RP2040 Zero:

![image](https://github.com/ByteWelder/Decktility/assets/3512789/4bb0c094-f211-4724-9735-7cd19dc698f1)

A bracket will be placed below the MCU to clamp it down.

**Comment by vault-thirteen** [comment]
Your project is very interesting !
